src/theory/strings/core_solver.cpp

```diff
@@ -1014,425 +1014,513 @@ void CoreSolver::processNEqc(std::vector<NormalForm>& normal_forms)
-                    info.d_ant.push_back( xgtz );
-                  } else {
-                    info.d_antn.push_back( xgtz );
-                  }
-                }
-                Node sk = d_skCache.mkSkolemCached(
-                    nfiv[index],
-                    nfjv[index],
-                    isRev ? SkolemCache::SK_ID_V_SPT_REV
-                          : SkolemCache::SK_ID_V_SPT,
-                    "v_spt");
-                // must add length requirement
-                info.d_new_skolem[LENGTH_GEQ_ONE].push_back(sk);
-                Node eq1 = nfiv[index].eqNode(
-                    isRev ? utils::mkNConcat(sk, nfjv[index])
-                          : utils::mkNConcat(nfjv[index], sk));
-                Node eq2 = nfjv[index].eqNode(
-                    isRev ? utils::mkNConcat(sk, nfiv[index])
-                          : utils::mkNConcat(nfiv[index], sk));
-
-                if( lentTestSuccess!=-1 ){
-                  info.d_antn.push_back( lentTestExp );
-                  info.d_conc = lentTestSuccess==0 ? eq1 : eq2;
-                  info.d_id = INFER_SSPLIT_VAR_PROP;
-                  info_valid = true;
-                }else{
-                  Node ldeq = NodeManager::currentNM()->mkNode( kind::EQUAL, length_term_i, length_term_j ).negate();
-                  if( ee->areDisequal( length_term_i, length_term_j, true ) ){
-                    info.d_ant.push_back( ldeq );
-                  }else{
-                    info.d_antn.push_back(ldeq);
-                  }
-                  //set info
-                  info.d_conc = NodeManager::currentNM()->mkNode( kind::OR, eq1, eq2 );
-                  info.d_id = INFER_SSPLIT_VAR;
-                  info_valid = true;
-                }
-              }
-            }
-          }
-          if( info_valid ){
-            pinfer.push_back( info );
-            Assert(!success);
+        // Since non of the other inferences apply, we just infer that `nc` has
+        // to start with the first character of the constant.
+        //
+        // E.g. "abc" ++ ... = nc ++ ... ---> nc = "a" ++ k
+        Node firstChar = stra.size() == 1 ? const_str
+                                          : nm->mkConst(isRev ? stra.suffix(1)
+                                                              : stra.prefix(1));
+        Node sk = d_skCache.mkSkolemCached(
+            nc,
+            isRev ? SkolemCache::SK_ID_VC_SPT_REV : SkolemCache::SK_ID_VC_SPT,
+            "c_spt");
+        Trace("strings-csp")
+            << "Const Split: " << firstChar << " is removed from " << const_str
+            << " (serial) " << std::endl;
+        info.d_conc = nc.eqNode(isRev ? utils::mkNConcat(sk, firstChar)
+                                      : utils::mkNConcat(firstChar, sk));
+        info.d_new_skolem[LENGTH_SPLIT].push_back(sk);
+        info.d_id = INFER_SSPLIT_CST;
+        pinfer.push_back(info);
+        break;
+      }
+      else
+      {
+        break;
```

Comment:
Can you comment what this break corresponds to?

Reply:
This is a bit tricky. It corresponds to the case where `isRev` is `true`. In this case, we would skip the following block in the original code: https://github.com/CVC4/CVC4/blob/499f1c70dbc869e2b1dd5f8a7b98d45bee2d29cc/src/theory/strings/core_solver.cpp#L1252-L1353 . After skipping this block, we essentially end up at the outer while loop and break out of it because `success` is `false`. In the refactored code, we have to break because we don't want to do a V-Split when `isRev` is `true`.

Reply:
Ah I see, can you add a comment to the code that says this?  I do remember this was the case but it looks a bit scary when we break without an inference.

Reply:
I've added the comment, PTAL.